prog/rand.go

```diff
@@ -754,3 +761,89 @@ func (a *CsumType) generate(r *randGen, s *state) (arg Arg, calls []*Call) {
 	// Filled at runtime by executor.
 	return MakeConstArg(a, 0), nil
 }
+
+// Finds a compatible resource with the type `t` and the calls that initialize that resource.
+func resourceCentric(t *ResourceType, s *state, r *randGen) (resource *ResultArg, calls []*Call) {
+	var p *Prog
+	for idx := range r.Perm(len(s.corpus)) {
+		p = s.corpus[idx].Clone()
+		resources := getCompatibleResources(p, t.TypeName, r)
+		if len(resources) > 0 {
+			resource = resources[r.Intn(len(resources))]
+			break
+		}
+	}
+
+	// No compatible resource was found.
+	if resource == nil {
+		return nil, nil
+	}
+
+	// Set that keeps the related references with the current resource.
+	relatedRefs := make(map[*ResultArg]bool)
+	for res := range resource.uses {
+		relatedRefs[res] = true
```

Comment:
This looks overly complex, I think this code can be simplified significantly if we store the resource itself in relatedRefs, and then in the loop below we also store a.Res instead of a. Then the checks in the loop below can be reduced simply to `relatedRefs[a] || relatedRefs[a.Res]`.

Reply:
Fixed.